src/main/scala/pipedsl/common/Utilities.scala

```diff
@@ -305,6 +307,167 @@ object Utilities {
     }
   }
 
+
+  def opt_func[A, B](f :A => B) : Option[A] => Option[B] =
+    {
+      case Some(value) => Some(f(value))
+      case None => None
+    }
+
+  /**
+   * Maps the function f_opt over the types of e1.
+   * MODIFIES THE TYPE OF e1
+   * @param e1 the expression to map over
+   * @param f_opt the function to apply to the types
+   * @return the expression with new types
+   */
+  private def typeMapExpr(e1 :Expr, f_opt : Option[Type] => Option[Type]) : Expr =
+    {
+      try
+        {
+          e1.typ = f_opt(e1.typ)
+        } catch
+        {
+          case _ :scala.MatchError =>
```

Comment:
This is kind of gross (hoping that `f_opt` throws scala.MatchError). We should properly use checked exceptions by giving f_opt a type of exception that we know it can throw and catching that.

Reply:
fixed by defining an option type that supports Some(value), None, and Error